src/expr/src/vector_op/agg/approx_count_distinct.rs

```diff
@@ -0,0 +1,254 @@
+            registers: [0; INDICES],
+        }
+    }
+
+    /// Adds the count of the datum's hash into the register, if it is greater than the existing
+    /// count at the register
+    fn add_datum(&mut self, datum_ref: DatumRef) {
+        if datum_ref.is_none() {
+            return;
+        }
+
+        let scalar_impl = datum_ref.unwrap().into_scalar_impl();
+        let hash = self.get_hash(scalar_impl);
+
+        let index = (hash as usize) & (INDICES - 1); // Index is based on last few bits
+        let count = self.count_hash(hash);
+
+        if count > self.registers[index] {
+            self.registers[index] = count;
+        }
+    }
+
+    /// Calculate the hash of the `scalar_impl` using Rust's default hasher
+    /// Perhaps a different hash like Murmur2 could be used instead for optimization?
+    fn get_hash(&self, scalar_impl: ScalarImpl) -> u64 {
+        let mut hasher = DefaultHasher::new();
+        scalar_impl.hash(&mut hasher);
+        hasher.finish()
+    }
+
+    /// Counts the number of trailing zeroes plus 1 in the non-index bits of the hash
+    fn count_hash(&self, mut hash: u64) -> u8 {
+        let mut count = 1;
+
+        hash >>= INDEX_BITS; // Ignore bits used as index for the hash
+        hash |= 1 << COUNT_BITS; // To allow hash to terminate if it is all 0s
+
+        while hash & 1 == 0 {
+            count += 1;
+            hash >>= 1;
+        }
+
+        count
+    }
+
+    /// Calculates the bias-corrected harmonic mean of the registers to get the approximate count
+    fn calculate_result(&self) -> i64 {
+        // Approximation for bias correction. See "HyperLogLog: the analysis of a near-optimal
+        // cardinality estimation algorithm" by Philippe Flajolet et al.
+        let bias_correction = 0.72134;
+        let m = INDICES as f64;
+        let mut mean = 0.0;
+
+        // Get harmonic mean of all the counts in results
+        for count in self.registers.iter() {
+            mean += 1.0 / ((1 << *count) as f64);
+        }
+
+        let raw_estimate = bias_correction * m * m / mean;
+
+        // If raw_estimate is not much bigger than m and some registers have value 0, set answer to
+        // m * log(m/V) where V is the number of registers with value 0
+        let answer = if raw_estimate <= 2.5 * m {
+            let mut zero_registers: f64 = 0.0;
+            for i in self.registers.iter() {
+                if *i == 0 {
+                    zero_registers += 1.0;
+                }
+            }
+
+            if zero_registers == 0.0 {
+                raw_estimate
+            } else {
+                m * (m.log2() - (zero_registers.log2()))
+            }
+        } else {
+            raw_estimate
+        };
+
+        answer as i64
+    }
+}
+
+impl Aggregator for ApproxCountDistinct {
+    fn return_type(&self) -> DataType {
+        self.return_type.clone()
+    }
+
+    fn update_with_row(&mut self, input: &DataChunk, row_id: usize) -> Result<()> {
```

Comment:
(Not related to this PR...) Why these functions of `Aggregator` returns `Result`? 😇